### PR TITLE
Elizabeth/scale-array

### PR DESCRIPTION
**Pull request by noot**
## Changes
Implemented vectors (slices, arrays) for scale encoding and decoding.  This is tested for []int, []bool, []*big.Int (the types we encode and decode)  Might need to do arrays of structs in the future.

## Tests:
```
go test ./codec/...
```

## Closes
ChainSafeSystems/go-pre#36

### CODE REVIEW

File: codec/decode.go [hunk 1]
```diff
@@ -229,3 +247,59 @@ func (sd *Decoder) DecodeTuple(t interface{}) (interface{}, error) {
 
 	return t, err
 }
+
+// DecodeIntVector decodes a byte array to an array of ints
+func (sd *Decoder) DecodeIntVector() ([]int, error) {
```
Comment (ansermino) [inline codec/decode.go]: Rename DecodeIntArray

File: codec/decode.go [hunk 2]
```diff
@@ -229,3 +247,59 @@ func (sd *Decoder) DecodeTuple(t interface{}) (interface{}, error) {
...
+	length, err := sd.DecodeInteger()
+	if err != nil {
+		return nil, err
+	}
+
+	sl := make([]int, length)
```
Comment (ansermino) [inline codec/decode.go]: Use `o` instead?

File: codec/decode.go [hunk 3]
```diff
@@ -229,3 +247,59 @@ func (sd *Decoder) DecodeTuple(t interface{}) (interface{}, error) {
...
+func (sd *Decoder) DecodeBoolVector() ([]bool, error) {
+	length, err := sd.DecodeInteger()
+	if err != nil {
+		return nil, err
+	}
+
+	sl := make([]bool, length)
+	for i := range sl {
+		sl[i], err = sd.DecodeBool()
+		//sl[i] = int(t)
```
Comment (ansermino) [inline codec/decode.go]: Remove comment

File: codec/encode.go [hunk 4]
```diff
@@ -167,3 +174,61 @@ func (se *Encoder) encodeTuple(t interface{}) (bytesEncoded int, err error) {
...
+
+// encodeVector encodes an interface where the underlying type is an array or slice
+// it writes the encoded length of the vector to the Encoder, then encodes and writes each value in the vector
+func (se *Encoder) encodeVector(t interface{}) (bytesEncoded int, err error) {
+	var n int
+	switch arr := t.(type) {
+	case []int:
+		n, err = se.encodeIntegerElements(arr)
+		bytesEncoded += n
+	case []*big.Int:
```
Comment (ansermino) [inline codec/encode.go]: The four following cases all use the same code except for the call to `se.encode...`

Comment (noot) [reply codec/encode.go]: unfortunately it must be this way due to the type switching. `arr` gets casted as the type in the case, if we put them all into one case it says `arr` is still an interface and doesn't compile

File: codec/decode.go (hunk 1 above)
Comment (noot) [reply codec/decode.go]: done

**Pull request merged**